### ISSUE TITLE
Kotlin Scripting Support

### ISSUE THREAD

**Issue by ligee** [Kotlin/KEEP#75]
https://github.com/Kotlin/KEEP/blob/scripting/proposals/scripting-support.md

replaces https://github.com/Kotlin/KEEP/blob/master/proposals/script-definition-template.md

**Comment by DaanDeMeyer** [comment]
This looks really promising!

Would this allow me to distribute a single self contained script template which could then be opened in Intellij without any other configuration being required? Ideally I could distribute a script template with a reference to the script definition and Intellij would automatically download this definition from Maven and provide rich code completion.

**Comment by ligee** [comment]
@DaanDeMeyer this is not yet in this KEEP, but that could be an interesting addition. I'll think about it. Thanks!

**Comment by artem-zinnatullin** [comment]
@ligee 2 questions:

* How is Spek related to scripting?
* What about unit testing for Kotlin Scripts? (tests for scripts themselves)?

Thanks!

**Comment by holgerbrandl** [comment]
Great that you guys are planning to push scripting to a new level. It's an amazing KEEP and I especially liked the bits about declarative bootstrapping of an IDE dependency  and runtime environment. It's also great to read how you envision to cover the whole range from build-files to shell scripts with a unified framework addition.

Though, it reads like a very ambitious plan to me including lots of sub-projects. So I'd be curious about the indented time-frame.


1. You mentioned the difficulties of parameterized shebangs, and indeed they don't work across shells, so using a declarative approach is the way to go I think.


2. It would be helpful to spec out the declarations first, so that third party tools like my kscript (see https://github.com/holgerbrandl/kscript) could support these declarations rather soon in order to fill the gap until an official JB solutions is ready. Once the latter is done, kscript would be obsolete for good.

Looking forward to kontlinconf so that we can chat more about it!

**Comment by ligee** [comment]
@artem-zinnatullin:
About Spec - it is described here - https://github.com/Kotlin/KEEP/blob/scripting/proposals/scripting-support.md#script-based-dsl. Basically, it means using scripting infrastructure to produce regular project class files from script-like sources with reduced boilerplate, like necessity to place spec tests inside a class.
About unit testing scripts - it is, of course, possible to use script runner inside tests, but we do not plan any additional helpers/facilities to simplify scripts testing. Maybe we'll consider it later though.

**Comment by ligee** [comment]
@holgerbrandl, thank you for the feedback. We will try to keep the size of the work here under control. Significant parts of the infrastructure changes described here are, in fact, reshapings and unifyings of the existing scripting infrastructure. But there should be, of course, some particular implementations on top of that, and implementing them will take time, but with clearly defined interfaces we may be able to outsource it partly to the community.
But at the moment it is work in progress, so the plans could change.
As of the specs for declarations - we'll try to keep it mind. But first, we have to finish more general infrastructure specs.

**Comment by artem-zinnatullin** [comment]
@ligee 

>About Spec - it is described here - https://github.com/Kotlin/KEEP/blob/scripting/proposals/scripting-support.md#script-based-dsl

Yep, I saw Spek there, that's why I asked :)

>like necessity to place spec tests inside a class

Spek is a pretty complex test framework that requires test engine/runner to work (currently Spek uses JUnit5 infrastructure). Also, usually you run tests through build system, not sure how scripting will handle that.

This can save one nesting level by removing top-level class (which doesn't look as a pain point atm), but on the other hand running tests seems to become much more complicated. Technically everything is possible, but I don't think this is direction we want to move Spek into at the moment.

cc @raniejade, @hhariri PTAL

**Comment by apatrida** [comment]
@ligee for script support, it looks like you have something roughed in for whitelist/blacklisting of functions/classes that can be accessed.  This will be important for us to embed Kotlin scripting in Elasticsearch or other engines like Arango, Neo4j, etc.  Ensuring what can be called at any given moment, or even what classes can be loaded.  Lastly, time limiting of the script running, or loop iteration limiting protections as well (endless loops kill these type of systems).  The more integrated this protection can be, the better.

**Comment by SolaKun** [comment]
![image](https://user-images.githubusercontent.com/3632663/30895219-a2775a66-a2fd-11e7-9cdc-3a00ba26b8b6.png)

**Comment by gtnarg** [comment]
Is there anywhere to follow the progress on this?

**Comment by ligee** [comment]
I'm working now on the updated proposal, along with the series of the prototypes. As soon as it will be ready, I'll publish the updated version. Hopefully, it will happen within a few weeks from now.

**Comment by sdeleuze** [comment]
I am interested to leverage these improvements in Spring and [Riff](https://github.com/projectriff/riff/issues/283).

**Comment by apatrida** [comment]
@ilya-g please add the requirement:  Between compilation and class loading, there should be the option for a verifier that can approve or reject a script.  This allows secure scripting to be tied into anything that executes those steps concurrently without needing to re-write every permutation of script runners that might be created.  Allowing something like Cuerentena integration https://github.com/kohesive/cuarentena ...   There should never be a hard-coded "compile and load class" without the ability to interject between those steps, before class loading.

**Comment by bhtek** [comment]
@apatrida Thanks. In cases where the script content came from outside of development, e.g. some simple formula for calculating P&L etc. It is critical to validate before loading for the feature to be usable in production.

**Comment by ligee** [comment]
The 1.2.50 is out with the new experimental scripting support. The KEEP is updated too, reflects now the approach taken with 1.2.50 and actual implementation status.
Not all feedback is incorporated (yet) into the KEEP, but I believe that what we have now is a good foundation that could be developed in the right direction relatively quick and easy.
Any feedback will be highly appreciated!

**Comment by ligee** [comment]
@apatrida it could be done now exactly at the moment you described, but having an extension point for it in the standard host interface looks like a good idea, I'll think about it.

**Comment by udalov** [comment]
@ligee As noted by @rgoldberg in #124, `kotlin-scripting-common` is listed twice in the "Implementation status" section, which is probably a typo

**Comment by yschimke** [comment]
It would be great if this optionally supported continuations via an implicit surrounding runBlocking call. Avoid nasty boilerplate and allow suspending functions to be top level.

**Comment by LouisCAD** [comment]
@cy6erGn0m You can already have top level `suspend fun` in `.kts` files

**Comment by yschimke** [comment]
@LouisCAD to clarify, I really meant calling them at the toplevel e.g.

https://github.com/yschimke/oksocial/blob/master/src/test/kotlin/commands/text-to-speech.kts#L71-L78

**Comment by ligee** [comment]
@cmorriss, seems a bug, thanks for pointing out! It was lost during the last refactoring when the source was removed from the configuration.

**Comment by casz** [comment]
I haven't found any good examples of how to actually setup a script project while having full IDE support and make it run. 😢

**Comment by cmorriss** [comment]
I've noticed that this seems to require access to the full kotlin compiler. Our use case is to run it as an embedded component, allowing scripts to be incorporated as sandboxed plugins in a backend service. It would be great if this could run utilizing kotlin-compiler-embedded as a dependency. It pulls in fewer dependencies and would allow this to more easily run as a library.

**Comment by cmorriss** [comment]
Just to add more context, when I try to run using the embedded compiler, I get the following exception:

java.lang.NoClassDefFoundError: com/intellij/openapi/util/Disposer
	at kotlin.script.experimental.jvmhost.impl.KJVMCompilerImpl.compile(KJVMCompilerImpl.kt:99)
	at kotlin.script.experimental.jvm.JvmScriptCompiler.compile$suspendImpl(jvmScriptCompilation.kt:43)
	at kotlin.script.experimental.jvm.JvmScriptCompiler.compile(jvmScriptCompilation.kt)
	at com.amazon.fuselets.kotlinscript.FuseletScriptCompiler.compile(FuseletScriptCompiler.kt:19)
	at com.amazon.fuselets.kotlinscript.FuseletScriptingHost$eval$1.doResume(FuseletScriptingHost.kt:28)
	at com.amazon.fuselets.kotlinscript.FuseletScriptingHost$eval$1.invoke(FuseletScriptingHost.kt)
	at com.amazon.fuselets.kotlinscript.FuseletScriptingHost$eval$1.invoke(FuseletScriptingHost.kt:14)
	at kotlin.script.experimental.host.BasicScriptingHost$runInCoroutineContext$1.doResume(BasicScriptingHost.kt:30)

It looks like this class, among others, has a modified package name (I'm guessing to prevent collisions when running as an embedded component) of org.jetbrains.kotlin.com.intellij.openapi.util instead of the normal com.intellij.openapi.util. Maybe these classes can be encapsulated and have the system catch the first ClassNotFoundException and pick the right one under the covers for future calls and instance creation needs. Just a thought.

**Comment by ligee** [comment]
@cmorriss In fact, there is no difference in size and dependencies between "full" compiler and embeddable one. The latter is intended for usages in the environments already containing some of the libs packed into the jar, e.g. IDEA's openapi. And it also requires the other components to be remapped similarly, that's why you're getting the error above. We do not provide yet the appropriate scripting components for using with the embeddable compiler (and I'm not sure we want to do it, taking into account related problems and misunderstandings).
In short, I do not recommend to do it. Please use the "full" compiler instead.

**Comment by cmorriss** [comment]
@ligee Thanks for the response. While it is possible to use the full compiler, we'll have to figure out a way to do so in a contained classloader as our particular use case does fall under the "environments already containing some of the libs packed into the jar" scenario.

In particular, this goes beyond just things like Google Guava. It actually includes the Kotlin standard libraries themselves. Our use case is to utilize Kotlin scripting as a backend service plugin language running in a tightly contained sandbox. This allows the logic of our services to be tweaked through these scripts by providing functional business logic in well defined extension points. 

These services all depend on Kotlin at a 1.2.x or 1.3.x level instead of a specific version, like 1.2.60. This allows for backwards compatible updates to be deployed without the need to update a ton of version numbers throughout our infrastructure. However, the compiler is clearly tied to a very specific version of Kotlin.

The problem is that by including the full compiler as part of our plugin library, it can cause conflicts as the Kotlin libraries that the service depends on may diverge from the version that is included in the compiler since they can be deployed and updated independently.

The only way we can avoid this issue currently would be to load the compiler in a separate classloader, perform the compilation there, and then run the classes in the classloader of the primary service. This can work (I hope!) as we can at least be sure that the Kotlin compiler is at or below the version of the Kotlin standard library for the service. We just can't be sure they're at exactly the same version level.

I believe that running with the embeddable version of the Kotlin compiler would solve these issues as it's embedded libraries would not conflict with those of the service embedding it.

In the mean time I'll work on the classloading solution, but would greatly appreciate if using the embeddable version of the compiler could be investigated further. I'd wager we're not the only use case for this, but I could be wrong.

Would be happy to hear I'm wrong about this as well!!!!

**Comment by jerves** [comment]
Hello, is anyone can tell me how can I define my kotlin script that has some functions with parameters, and how can I visit the functions in my custom project?  With the new experimental scripting support.

**Comment by cmorriss** [comment]
@jerves 
Not sure how "correct" this is, but it works.

https://github.com/cmorriss/kotlinscript-sandbox/blob/master/src/main/kotlin/KotlinScriptSandbox.kt

Essentially, the script has a function that matches the signature of a function in the abstract script class. The script is compiled to a class file, instantiated, and cast to the abstract script class.

At that point, you can call the function, passing in whatever parameters are needed, and it will execute the code in the script and give you back a return value.

**Comment by cmorriss** [comment]
@ligee Thanks for the reply. Good to know. I definitely want to utilize a supported path and iterating over the functions and using reflect can achieve the same result.

It would be a "nice to have" feature to supply an interface to the scripting host, in addition to the script source, such that it would return an instance of that interface that could be called to execute corresponding functions in the script. If the script does not provide functions that match the signatures of the interface, an exception is thrown.

I'm planning to provide this using a dynamic proxy to forward the calls to the script instance functions, although it would be very convenient for others who have a similar use case be able to access some similar functionality that's baked into the scripting support.

**Comment by ligee** [comment]
@cmorriss I created an issue to track this topic - https://youtrack.jetbrains.com/issue/KT-27349

**Comment by cmorriss** [comment]
Thanks @ligee . I also created a separate issue to track the possibility of creating an embeddable version of the jvm scripting host that would work with the embeddable compiler here  - https://youtrack.jetbrains.com/issue/IDEA-199942

**Comment by jerves** [comment]
> @cmorriss I created an issue to track this topic - https://youtrack.jetbrains.com/issue/KT-27349

Thanks！It is exactly my use case. And I want to know in which version can support this use case?

**Comment by apatrida** [comment]
> @apatrida it could be done now exactly at the moment you described, but having an extension point for it in the standard host interface looks like a good idea, I'll think about it.

Did this happen as described @ligee , I'm about to dig in to update my scripting libs

**Comment by ligee** [comment]
@apatrida I haven't implemented any extension points yet to simplify such checks. Please try to make a custom host for that and tell me if there is anything missing now. Alternatively, there is a compiler cache interface that receives the compiled bytes, so maybe it is possible to hack something there. And finally, you can try to wrap the default script compiler implementation and make checks there.
I'll try to add missing pieces fur such functionality on the next API update iteration.

**Comment by NikkyAI** [comment]
Scripting support so far is awesome, good job to all parties involved

The Feature i think i need next would be `import`/`include` for extra source files without compiling them into a jar in a seperate step (because thats a hell of a lot of complication and things to go wrong)
The files I want to include contain constants generated at setup by a gradle plugin/task in the project source, i am aiming for somwhat of a moving target so it cannot be compiled into the library
one solution i found that works.. (for now) is to concat all the files i need into one big string and move imports to the top with regex
needless to say this is not the prettiest code or easy to reason aboutor easy to track down compile errors since line numbers will not even match up anymore

and thats about all the complaints i have i guess.. anything else this api handles or will be bale to handle with ease it seems

**Comment by ligee** [comment]
@balage1551, the compiling from a string is supported from the beginning. The restrictions are planned, but the timeline is not yet clear.

**Comment by abelsromero** [comment]
What's the status of implemented features? I don't mind dealing with beta and things that may change but I am finding the documentation is not clear, and all the examples I have found don't seem to work.
For instance, I am trying to run the example here https://blog.jetbrains.com/kotlin/2018/09/kotlin-1-3-rc-is-here-migrate-your-coroutines/ and the one from https://github.com/Kotlin/KEEP/blob/scripting/proposals/scripting-support.md
```
@file:DependsOn("junit:junit:4.11")

org.junit.Assert.assertTrue(true)

println("Hello, World!")
```

But the resolution simply does not seem to work.

**Comment by ligee** [comment]
@abelsromero there is a section on the status in the KEEP, it might be not 100% accurate, but you can get a picture.
The example from the blog post definitely works, I checked it many times. Possible problems might be either you're not pulling the right script definition (e.g. do not add `kotlin-main-kts.jar` to the classpath), or the script "type" is not recognized (e.g. in the example from the blogpost, you have to give your script the extension ".main.kts", this is the extension by which `kotlin-main-kts` recognizes its scripts).
Unfortunately, the diagnostics and proper tutorials and examples are lacking right now, we're working on it.

**Comment by dobrakmato** [comment]
Will this feature be supported on Kotlin/Native or it is planned to be JVM only?

**Comment by abelsromero** [comment]
Thanks @ligee, totally my fault for rushing to copy-pasting the examples without paying attention. I got it working now! Keep up the good work!
Those stuck with Windows are looking forward to have a standard scripting tools.

**Comment by ligee** [comment]
@dobrakmato the API is designed to be platform-agnostic as far as possible, but current implementations are JVM-only. We're investigating possibilities to make it available in Kotlin/Native too, but it is too early for conclusions.

**Comment by SillyMoo** [comment]
Any tips on how to get intellij code completion working for scripts? I'm able to dynamically load files, and all works well, but code completion is not working (as intellij can't seem to resolve the base class from the ```@KotlinScript``` annotation.
I get the impression it has something to do with the intellij kotlin compiler properties "Script template classes", but I can't figure out how.

**Comment by ligee** [comment]
@SillyMoo there are several possibilities to supply Intellij with everything needed for the resolution/completion/etc. 
The simplest one is, as you mentioned, kotlin compiler properties "Kotlin Scripting". You need to put the FQN of the base class into "Script Template Classes" field and the classpath needed to load this class into "Script templates classpath". These settings are intended to be more like workaround though.
If your script is a part of the regular project sources, then you just need to add a properly prepared jar with the script definition into the project dependencies.
And finally, it is possible to write a simple Intellij plugin that will supply needed info for your scripts. This is how e.g. "build.gradle.kts" files are supported now.
We're also considering other methods of configuring Intellij for scripting, but so far haven't come up with good enough solutions.

**Comment by SillyMoo** [comment]
@ligee, so I have an uber jar with the ```@KotlinScript``` annotated script class in it, and also a file in /META-INF/kotlin/script/templates with the name consisting of the FQDN of the annotated script class. But intellij does not seem to recognise my script file extension (it does do so in the original project from in which the scripting class is generated).
Is there any debug I can enable to see if there is a parsing issue or something?

**Comment by SillyMoo** [comment]
Found the issue, I was bundling some runtime components in with the shaded Jar by mistake, this seems to have caused some class loading issues.

**Comment by alshain** [comment]
Regarding the discovery files: I realize that the content of these files is irrelevant/they should be empty, but why not add a *.txt extension?

When I use "Search Everywhere" to navigate to the file, I often habitually just hit "Enter", then IntelliJ wants me to associate it with a file type, because it's trying to open it.

OTOH, by not having an extension, *maybe* it's clearer that the file itself shouldn't be edited?

**Comment by ligee** [comment]
@alshain It seems to me that ".txt" would be misleading. But I understand the problem with the association, maybe there is some way around it. I noted it for future consideration - https://youtrack.jetbrains.com/issue/KT-28593

**Comment by alshain** [comment]
@ligee Thank your for the quick response and thanks for considering.
Some other ideas in case it does end up becoming a file extension: *.empty, *.classname, *.ktsdiscovery

**Comment by unlimitedsola** [comment]
the default `Run Scratch File`  (`Ctrl+Alt+W` by default on Windows/Linux) keyboard shortcut makes me so embarrassing.

**Comment by NataliaUkhorskaya** [comment]
@unlimitedsola The main reason is that `Ctrl+Enter` is already assigned to `Split lines` Action. 
The better place to discuss it is https://youtrack.jetbrains.com/issue/KT-24180

**Comment by bernaferrari** [comment]
Hey! It comes to my surprise **no one ever** talked about clipboard support. Where is it? I want to read/write on it. On most places, [people use AWT, some JavaFX, some even xclip](https://stackoverflow.com/questions/3591945/copying-to-the-clipboard-in-java) (which requires unix and installing it first). I think scripts could benefit a lot if we had this.

**Comment by Aivean** [comment]
Hi, I followed the example here: https://github.com/Kotlin/KEEP/blob/scripting/proposals/scripting-support.md#examples  and it works, but one of my scripts fails compilation with the error:

     ScriptDiagnostic(message=Cannot inline bytecode built with JVM target 1.8 into bytecode that is being built with JVM target 1.6. Please specify proper '-jvm-target' option, severity=ERROR, location=Location(start=Position(line=109, col=13, absolutePos=null), end=null), exception=null)

I don't understand what is happening. Does this mean that script is compiled to 1.6 target for some reason? My "host" project is compiled to 1.8 target.

I tried adding `-jvm-target 1.8` via `kotlin.script.experimental.api.ScriptCompilationKt#getCompilerOptions`, but it doesn't seem to do anything. I also found an annotation `kotlin.script.templates.ScriptTemplateAdditionalCompilerArguments`, but I'm not sure how to use it.

Any ideas? 
Thanks.

**Comment by cmorriss** [comment]
Assuming you're using Gradle, make sure this is in your build.gradle file:

```
tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
  kotlinOptions {
    jvmTarget = "1.8"
  }
}
```

**Comment by Aivean** [comment]
> Assuming you're using Gradle, make sure this is in your build.gradle file:
> 
> ```
> tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
>   kotlinOptions {
>     jvmTarget = "1.8"
>   }
> }
> ```

I already have 
```
compileKotlin {
  kotlinOptions.jvmTarget = "1.8"
}
```
in my single-module project, but I'll try your suggestion.   Thanks.

However, I think this has something to do with JVM target for scripting engine, not the "host" project.

**Comment by NikkyAI** [comment]
the problem is that the host project is compiled for target 1.8 and kotlin scripting atm ignores those options.. and uses the default of 1.6
but it is possible to work around that

import the file you need to include with `importScripts.append(File(somePath).toScriptSource())`
in the `jvm` block
this is available in 1.3.20-eap-100 or 1.3.30-dev versions

about the jvm-target stuff i think its this:
https://youtrack.jetbrains.com/issue/KT-27815

**Comment by ligee** [comment]
@Aivean currently setting the compiler options for script compilation is not possible due to the 
https://youtrack.jetbrains.com/issue/KT-27815. Unfortunately, it will not be fixed in the upcoming 1.3.20.
But I think for 1.3.20 I can change the default to 1.8, it should help in most situations. I created https://youtrack.jetbrains.com/issue/KT-29319 to track this issue.

**Comment by Aivean** [comment]
@NikkyAI, @ligee thanks for the help. It makes sense now. Subscribed to both issues.

**Comment by NikkyAI** [comment]
is this also the right place to chat about limitations in the current implementation?

i found that any Annotation in scripts that contain any kind of array
eg. `val categoryIds: IntArray = []` or `val versions: Array<String>`
are only available as `InvalidScriptResolverAnnotation`
in `beforeCompililation` or in `onAnnotations`
it also seems to dislike using single enum values

**Comment by efemoney** [comment]
@NikkyAI I had the same issue. There's a bug somewhere in the script compiler. That `InvalidScriptResolverAnnotation` signifies an error. Somewhere in [this method](https://github.com/JetBrains/kotlin/blob/a4d1a8f60addaa8151bbcd4b36ee53f97adae193/compiler/frontend.script/src/org/jetbrains/kotlin/script/scriptAnnotationsPreprocessing.kt#L39)

**Comment by NikkyAI** [comment]
also another thing.. annotations with default parameters and parameters left out work when executing the script but will result in `InvalidScriptResolverAnnotation` in idea

**Comment by ligee** [comment]
I think it is better to discuss implementation problems in slack and YT, and reserve this place for discussions about the proposal and general scripting concepts.

**Comment by pdvrieze** [comment]
I had a look at the proposal. I'm conceptually interested in the native case with a focus on fast initialisation (as a shell script). This would probably take shape of "compilation" being the creation of an in-memory representation of an AST (perhaps with some minor optimizations applied).

In that sense the interface of `CompiledScript` is not helpful as it returns a KClass. It appears that the API is not designed to be independent from the evaluation environment (this works in the JVM due to dynamic class loading). The type of "compilation result" should be completely implementation dependent and just require some interfaces.

The evaluation configuration is also quite limited. It would be expected that script types can inherit "native" types and use native functions, sometimes the other approach (native types implementing script types) may be worthwhile to allow for native plugins. It does not seem that the interface is thought through enough in this respect.

**Comment by utybo** [comment]
I cannot seem to find how to configure IntelliJ IDEA to provide code completion. For context, I am working in a Gradle project. I have the annotated class with `@KotlinScript` in the main folder, while my `kts` scripts are in the test folder. The code completion works like a regular Kotlin script file, and does not include any completion for members specific to the class annotated with `@KotlinScript`.

> The simplest one is, as you mentioned, kotlin compiler properties "Kotlin Scripting". You need to put the FQN of the base class into "Script Template Classes" field and the classpath needed to load this class into "Script templates classpath". These settings are intended to be more like workaround though.

I cannot find any option named "Script Template Classes" or "Script templates classpath", although there is a read-only list of the templates currently recognized by IntelliJ.

In the proposal:

> if the script definition module is in the same project as the usage part, the module (and in particular - base class) should be compiled to class files in order to be recognized by the discovery mechanism

I tried to build the project by using Build > Build Project but that did not seem to do anything...

My apologies if this is the wrong place to ask this.

**Comment by utybo** [comment]
Nevermind, I simply forgot to add the file in META-INF/kotlin/script/templates

This should probably be mentioned somewhere in the examples, more prominently than hidden away in a more technical section (though I understand that this proposal is not really a tutorial for newcomers)

**Comment by oshai** [comment]
@ligee can you give a brief update what are the plans with this keep? I lost track with few of the off-topics comments and it seems there haven't been a change here for a while...

**Comment by lare96** [comment]
I'm on 1.3.50 and no matter what I try, it seems like my script definition class never gets found on the classpath. Here is my script definition file

```kotlin
package api.predef

import kotlin.script.experimental.annotations.KotlinScript

@KotlinScript(
        displayName = "Luna plugin script",
        fileExtension = ".kts", // luna.kts
        compilationConfiguration = MyScriptCompilationConfiguration::class
)
abstract class LunaPluginScript
```

and I have the template file here (I've tried putting it in other places as well)

![](https://i.imgur.com/Rxe1d84.png)

**Comment by utybo** [comment]
@lare96 A few things, the name of the file in META-INF/kotlin/script/templates should be the full name of the annotated class (so just get rid of the `.kt` at the name as the full name would be `api.predef.LunaPluginScript`), and I think you have to give it a more specific file extension, like `luna.kts`, otherwise it might be trying to match your files against another script definition.

**Comment by lare96** [comment]
@utybo still doesn't seem to work. All scripts are being compiled with ScriptTemplateWithArgs as their superclass still. I'm running the server with gradle, do you know if there are any additional steps that need to be taken to get it to work?

**Comment by ligee** [comment]
@lare96 I think there are better places to ask help with this, namely #scripting channel on the Kotlin Slack. This thread is dedicated for the KEEP discussion.
But the short answer - with gradle you need to enable script definitions discovery explicitly, because discovery is too expensive in gradle case, see e.g. this test: https://github.com/JetBrains/kotlin/blob/master/libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/resources/testProject/scriptingCustomExtension/app/build.gradle#L5

**Comment by ligee** [comment]
@oshai I'm going to update the KEEP with the latest status in the near future. In short - there are no big changes, we continue to add features to the API and implementation, and intend to do it for some time more until we'll understand that it is good enough to bring it to the released status.

**Comment by tobia** [comment]
The shebang problem is well known. Ironically Linux, the most used OS among those that support a shebang line, is also the worst offender: to this day it splits the line on the first whitespace only.

There are two ways to get around it:

1. Use the new `-S` option recently added to the `env` command in Coreutils 8.30 (2018-07-01) but this will take years to propagate to distributions.
2. Write a "polyglot" script.

The second option is a time-honoured hack, or clever trick, that consists in writing a script that is syntactically valid in more than one language, usually `/bin/sh` (the POSIX shell) and the target language. In Kotlin and other languages that support `//` as a line comment, this is particularly easy to do:

```
#!/bin/sh
//bin/true; exec kotlinc -script "$0" "$@"

println("Hello, world!")
for (arg in args) println(" - '$arg'")
```

This should work on all POSIX systems IIRC. In case it's not obvious, this is what's happening:

1. The OS reads the shebang line and executes `/bin/sh`
2. The POSIX Shell runs the command `//bin/true`, which is just another way of writing `/bin/true`, which does nothing.
3. Then the shell `exec` (replaces its process with) the Kotlin compiler, passing it any options that are needed (`-script`, `-cp`, etc.) plus the location of the script file itself `"$0"` and any additional arguments that were provided on the command line `"$@"`
4. The Kotlin compiler skips the shebang line `#!` and the comment `//` and starts compiling and running the rest of the script.

This is an example execution that shows how argument whitespace is preserved:

```
$ ./test.kts 1 '2 3'
Hello, world!
 - '1'
 - '2 3'
```

If you need further initialization logic, such as exporting environment variables, you can open a Kotlin multiline comment by hiding `/*` inside a shell string:

```
#!/bin/sh
//bin/true '
/*'
export PLANET="World"
exec kotlinc -script "$0" "$@"
*/

println("Hello, ${System.getenv("PLANET")}!")
for (arg in args) println(" - '$arg'")
```

Or more simply, you can rely on the absence of filesystem matches for `/*bin/true` being lexicographically before `/bin/true` (which should hold true for most systems):

```
#!/bin/sh
/*bin/true
export PLANET="World"
exec kotlinc -script "$0" "$@"
*/

println("Hello, ${System.getenv("PLANET")}!")
for (arg in args) println(" - '$arg'")
```

I have no idea if any of these "hacks" have any place in the current discussion. I just thought I'd mention them in case somebody finds them useful.

**Comment by zhelenskiy** [comment]
@ligee Is there any table that shows the current status of developing Kotlin Scripting Support? It should be something like:

| Feature | Description | Status |
| :--: | -- |-- |
| Name | ... | ... |

If that exists, where can I find it?
If not, it would be great if it existed.

**Comment by Golvanig** [comment]
Hi,
Thanks for this new feature.
I tried the @file:Repository with "1.4-M2" 
as in "kotlin-main-kts-test/testData/kotlinx-html.main.kts"
And could not figure out how it works?
kotlin-main-kts-test/testData/kotlinx-html.main.kts is not working for me
Every other examples are OK

**Comment by ligee** [comment]
@zhelenskiy, the best approximation to this is the status part of this KEEP, which I'm updating from time to time. The next update is expected along with the 1.4.0 release.

**Comment by zhelenskiy** [comment]
@ligee I look forward to the update (1.4.0) then! Is there any information about the probable date of the release?

**Comment by ligee** [comment]
@Golvanig, what exactly doesn't work for you? The test you mentioned works for me.
You probably can find simpler to reuse examples in the repo https://github.com/Kotlin/kotlin-script-examples, although it is still based on the 1.3.70.

**Comment by Golvanig** [comment]
@ligee, Thanks for the feedback.
When I launch the script i.e.:
"kotlinc -script kotlinx-html.main.kts -cp /opt/kotlin/1.3.72/kotlinc/lib/kotlin-main-kts.jar"

I have the following results.
It seems that @file:repository is not found.
All the other examples work fine, for instance @file:DependsOn is OK.

---------------------------------------------------------
:: problems summary ::
:::: WARNINGS
		module not found: org.jetbrains.kotlinx#kotlinx-html-jvm;0.6.11

	==== central: tried

	  https://repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-html-jvm/0.6.11/kotlinx-html-jvm-0.6.11.pom

	  -- artifact org.jetbrains.kotlinx#kotlinx-html-jvm;0.6.11!kotlinx-html-jvm.jar:

	  https://repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-html-jvm/0.6.11/kotlinx-html-jvm-0.6.11.jar

error: unresolved reference: Repository (kotlinx-html.main.kts:3:7)
error: unresolved reference: DependsOn (kotlinx-html.main.kts:4:7)
error: unresolved reference: kotlinx (kotlinx-html.main.kts:6:8)
error: unresolved reference: kotlinx (kotlinx-html.main.kts:6:31)
error: unresolved reference: kotlinx (kotlinx-html.main.kts:6:61)
error: unresolved reference: createHTML (kotlinx-html.main.kts:10:7)
error: unresolved reference: body (kotlinx-html.main.kts:11:5)
error: unresolved reference: h1 (kotlinx-html.main.kts:12:9)
error: unresolved reference: + (kotlinx-html.main.kts:12:14)
kotlinx-html.main.kts: error: unresolved dependency: org.jetbrains.kotlinx#kotlinx-html-jvm;0.6.11: not found
kotlinx-html.main.kts:3:7: error: unresolved reference: Repository
@file:Repository("https://jcenter.bintray.com")
      ^
kotlinx-html.main.kts:4:7: error: unresolved reference: DependsOn
@file:DependsOn("org.jetbrains.kotlinx:kotlinx-html-jvm:0.6.11")
      ^
kotlinx-html.main.kts:6:8: error: unresolved reference: kotlinx
import kotlinx.html.*; import kotlinx.html.stream.*; import kotlinx.html.attributes.*
       ^
kotlinx-html.main.kts:6:31: error: unresolved reference: kotlinx
import kotlinx.html.*; import kotlinx.html.stream.*; import kotlinx.html.attributes.*
                              ^
kotlinx-html.main.kts:6:61: error: unresolved reference: kotlinx
import kotlinx.html.*; import kotlinx.html.stream.*; import kotlinx.html.attributes.*
                                                            ^
kotlinx-html.main.kts:10:7: error: unresolved reference: createHTML

**Comment by cortinico** [comment]
> It seems that @file:repository is not found.

That's a known issue
https://youtrack.jetbrains.com/issue/KT-37293

We had the same problem on detekt/detekt here https://github.com/detekt/detekt/pull/2777#issuecomment-640095142

**Comment by Golvanig** [comment]
Thanks @cortinico, 
Just for the sake of this awaiting feature:
The behavior looks different whether we use -script option or not.

**Comment by raniejade** [comment]
@ligee Are there plans to support Kotlin/Native? I can't imagine how the support would look like, embedding the compiler for every platform will be a PITA. Maybe we can get a way by just embedding the JVM?

**Comment by ligee** [comment]
@raniejade definitely not in the nearest future. We're investigating the possibilities to make it from time to time, but so far it looks too complicated to be useful and we do not found yet enough justification to put enough effort into it. But if you have a good scenario for it - please share, this may help us to understand the possible use cases better.

**Comment by raniejade** [comment]
@ligee Totally understandable, our immediate use case is scripting in Godot (https://github.com/utopia-rise/godot-kotlin). At the moment, we have to build a binary (dynamic library to be exact) whenever a change is made to a kotlin script. It is a bit of a pain as compiling down to a dynamic library takes time, so you can image how slow the fix-debug cycle can be. KN is used so that we can interact with Godot's c api directly, alternatively we could have just embedded the JVM and use JNI/JNA but we don't want to deal with the overhead of using those tech. I guess with Project Panama, that path becomes a bit more desirable. 

As a side note, I wonder would it be possible to add support a scripting host via a third party compiler plugin instead of the support being built within the compiler itself? Imagine being able to write JVM scripts in a KN host using the kotlin scripting api.

**Comment by ligee** [comment]
@raniejade, thanks for the info, we'll consider this scenario at the next planning phase (also cc: @SvyatoslavScherbina).

The Kotlin scripting is already a compiler plugin, although since we do not have a well-defined plugins API yet, it requires special ad-hoc handling in the compiler, and we're not yet ready to make it third-party hackable. But this time may come someday.

**Comment by xGnoSiSx** [comment]
Does the assorted machinery and kotlin compiler work on android and pass the bytecode to ART?
Or is this feature to be supported only on desktop? Just checking cause I found no relevant documentation for this. What I ask is if: kotlin-scripting-jvm and jvm-host support working under the ART. For reference, I was able to do this with groovy. 

My case study is sending kotlin scripts to a mobile device/tablet for compilation & execution. Thanks!

**Comment by pdvrieze** [comment]
I don't think there is a direct-to-art backend for the compiler. Android doesn't come with the class to art translator either so I don't think it would be supported on Android. Kotlin scripting isn't really like normal scripting in that it does actually compile the script to executable code, not merely as an optimization, so integration is harder (and startup cost is much higher even if long term speed is better).

**Comment by way-zer** [comment]
Can I compile kts to js in JVM platform using scripting?

**Comment by gildor** [comment]
> the class to art translator

On Android you don't need any art translation, art is a runtime, it's not even an API, it's a device implementation detail.
DES is the actual bytecode format

And dynamically load Java class files is completely possible on Android, see [DexClassLoader](https://developer.android.com/reference/dalvik/system/DexClassLoader)

**Comment by ligee** [comment]
> Can I compile kts to js in JVM platform using scripting?

There is a preliminary prototype of such compilation, but it is nowhere near completion yet.

**Comment by ligee** [comment]
> Does the assorted machinery and kotlin compiler work on android and pass the bytecode to ART?

It should theoretically be possible but seems there are quite some practical issues on the way, see e.g. https://youtrack.jetbrains.com/issue/KT-33560

**Comment by xGnoSiSx** [comment]
@ligee Hi, yes I'm aware of KT-33560 since I already posted on it. Thank you for your efforts and for the reply.

I'm a bit caught by surprise on this. I adopted kotlin on my commercial project for the end user scripting capabilities as well, replacing groovy, but perhaps I may have to step back a bit especially  if I plan to release on 2021.

Of course I'm not expecting this to do the work that ART does, just pass the byte code. You can take a look on the groovy source code for the exact details on this. No problems if you change the scripting api or add another one just for android - its experimentalxas it stands.

I'm willing to pay the price of compilation and warm up on the users device, even by displaying a spinner or progress bar, not asking for miracles here.

**Comment by BenWoodworth** [comment]
Has there been any talk about allowing `@file:Include`  to include script files from places other than the locale filesystem, e.g. from HTTP URLs? I think it would be nice have, since it would make posting things like utility scripts a lot easier (compared to publishing maven artifacts and using `@DependsOn`). Importing like this is already being used in other runtimes, [such as Deno](https://deno.land/manual/linking_to_external_code#linking-to-third-party-code) (the potential successor to Node.js)

I *could* see this being somewhat of a security risk (since it's a lot easier to add malicious code later on), but there are definitely ways to avoid that (e.g. optionally requiring the included file's source to match a specified hash code).

**Comment by zhelenskiy** [comment]
> Has there been any talk about allowing `@file:Include` to include script files from places other than the locale filesystem, e.g. from HTTP URLs? I think it would be nice have, since it would make posting things like utility scripts a lot easier (compared to publishing maven artifacts and using `@DependsOn`). Importing like this is already being used in other runtimes, [such as Deno](https://deno.land/manual/linking_to_external_code#linking-to-third-party-code) (the potential successor to Node.js)
> 
> I _could_ see this being somewhat of a security risk, introducing the possibility of man-in-the-middle attacks, and making it much easier to add malicious code later on, but there are definitely ways to avoid these vulnerabilities like requiring https:// or using hash-based file verification.

I don't see any risk there because you still can use java.policies to forbid some types of code sources.

**Comment by ligee** [comment]
@BenWoodworth, the support for non-file-based script sources is implemented in the API, but some implementation details prevent it from working right now for the `@Import` case. This is a known problem that we're going to address at some point. It seems it was never properly described, so I created an issue to track it - https://youtrack.jetbrains.com/issue/KT-42810

As for the security issues, it is probably up to the scripting host author to apply the necessary security measures. Most likely we will not support remote scripts in our default hosts out of the box, at least not by default.

**Comment by cromefire** [comment]
As the `restrictions` part isn't implemented yet, is there any planning for any guarantees on that?

My use case would be to use those as rule scripts and there for it would be quite beneficial if it was possible to make sure that the script has only ever access to certain classes and their public members, because that would allow the restriction of misuse by modifying the application in disallowed ways, create and delete files and much more. I understand that that isn't an easy job, but as it stands kotlin would be as far as I'm aware the only option that provides proper code completion, speed (using compilation & hit) and  sandbox capabilities.

I've seen that kotlin scripting available in TeamCity for example and I wonder does it do some kind of sandboxing already or is it just evaluating the script and assuming it's trusted, because if it isn't sandbox/restricted it could do potentially do anything inside TeamCity, right?

Other "solutions" wold be:
- Running without restrictions -> Attacks possible
- Using random scripting language xyz -> rarely the tooling for proper code completion, not statically typed, not compiled (not compiled), sometimes hard to integrate with the JVM
- Compiling kotlin to WASM and trying to integrate that -> Huge effort, probably lower performance, if it works at all
- Compiling kotlin to js and trying to integrate that -> Probably large resulting file, huge effort, slower, probably possible

With that in mind kotlin could probably enable a really easy (for users, not developers) way to have easy, fast and good (typesafe, real time code analysis, DSL capable & completions available) sandboxed scripting in the JVM.

**Comment by utybo** [comment]
> Edit: The JVM's ClassLoader and SecurityManager might be handy to use for an implementation as they seem to provide the primitives for sandboxing JVM code and would allow offloading of the guarantees to the JVM.

Using `SecurityManager` wouldn't work as it is deprecated for removal in Java 17 and will be removed at a later date (see [JEP 411](https://openjdk.java.net/jeps/411))

**Comment by ligee** [comment]
@cromefire the restrictions so far is only an idea, we are not yet started to do any related design. So no guarantees at the moment.
But you can go quite far at the moment with that using already provided customization possibilities - you can disable stdlib and reflec via compiler options, provide very strict compilation classpath and run your script in an isolated classloader (or provide your own sandboxing for the bytecode issued by the scripting compiler).
As of the TeamCity, if you're talking about the kotlin scripting build steps, then there is no dedicated sandboxing for those, as far as I know, but they are in the same company as any other build steps, that could be written in many scripting languages and do whatever they want. But as far as I can tell (and I'm definitely not an expert here), since they run on the build agents, this is not considered a threat, provided that you're following a proper security practices of bidirectional protection of your build infrastructure.

**Comment by xGnoSiSx** [comment]
To be fair, I gave up expecting progress on this. I can have more success if I sit down and code some version of lisp in kotlin over the weekend, as a solution.

**Comment by Snipx** [comment]
Hi, could anyone give me some context on how can I connect `ScriptEvaluationConfiguration` and `ScriptCompilationConfiguration`? Namely, I want to pass the state/data from `RefineScriptCompilationConfigurationHandler` to `ScriptEvaluationConfiguration` somehow.  The specific use case is setting `providedProperties` conditionally, based on presence of some annotation in a script. I am using `RefineScriptCompilationConfigurationHandler` to find the annotation of the given type in the `ScriptCollectedData` and based on whether I found that annotation, I add new property to `providedProperties` of the `ScriptCompilationConfiguration`. However, I also need to pass the property value into `ScriptEvaluationConfiguration` based on the very same condition and I cannot seem to find a way to do so (check whether the annotation is present in the source file). I have explored `ScriptEvaluationConfigurationRefinementContext` and I can't see a way to go back to the `ScriptCollectedData` or something like this. The closest I've got is `context.compiledScript.getClass(context.evaluationConfiguration).valueOrNull()?.annotations` but it returns empty list, probably because `@file:Abc` annotation does not end up in the compiled script.

I the documentation for `Scripting host configuration` I read that "These properties are defined by the scripting host and contain general parameters needed for all scripting services." but I don't see a way to store those options in the hosting config and it does not seem right at all.

Thanks. If the description is not clear I am happy to provide more context.

**Comment by ligee** [comment]
@Snipx, it seems that your scenario is not covered, and it makes sense to introduce a link from evaluation configuration to the compilation one. Could you please file the issue in YouTrack, basically with the description you provided.
As a workaround you can create a variable shared between configurations. See, e.g., how `providedProperties` property is defined, define your own similarly, make it pointing out to a separately allocated object from both configurations and modify this object in your handler.

**Comment by Snipx** [comment]
Leaving the link to the ticket for introducing the link from evaluation configuration to the compilation configuration: https://youtrack.jetbrains.com/issue/KT-48758

**Comment by aartiPl** [comment]
Hi, I am the current maintainer of 'kscript' and would like to align kscript with Kotlin Scripting implementation.
@ligee - what is the current status of Kotlin Scripting? Is there a list of annotations (with exact syntax) that, would be used in KS?

I also wonder if Kotlin Scripting can be used in kscript as backing implementation. Can you elaborate on that?

**Comment by ligee** [comment]
Hi, @aartiPl!
I'm glad to hear that you're thinking about using Kotlin Scripting API in kscript. A couple of years ago, I've tried to do some porting myself, but it ended up to be a bit more work than I anticipated. But it will be very nice if you'll do it, this will allow us to use common core implementation with all the relevant benefits of it.

The status of the relevant API is still "experimental", but we are planning to "deexperimentalize" it in the nearest future, maybe even in 1.8.x. This will mean that we'll move the API to the new packages without "experimental" part, and agree to support it in the foreseeable future. There are no significant changes planned in the API itself so far.

So I guess you can freely proceed with "porting" or aligning now, just take into account the future renaming.

There is very little documentation so far, unfortunately. The annotations and other part of the API could be extracted for the documented source code in the Kotlin repo. All relevant subprojects are located at https://github.com/JetBrains/kotlin/tree/master/libraries/scripting.

**Comment by aartiPl** [comment]
In regards to annotations - I can find only 3 in the source code:
![obraz](https://user-images.githubusercontent.com/1508798/186132128-692ec57c-0949-4b09-832e-e44e61067a5f.png)

But kscript is having much more:
MavenRepository, Repository -> (ok)
DependsOn, (DependsOnMaven) -> (ok)
Include -> can not find it; in some snippets, I could see @Import annotation
EntryPoint ->(?)
KotlinOpts, KotlinOptions -> (?)
CompilerOpts, CompilerOptions -> (?)

In the new kscript (4.1) I have deprecated all comment-based annotations. Also, I have changed existing annotations to be aligned with Kotlin-scripting annotations:
1) added Repository (in addition to deprecated MavenRepository)  -> (how to pass user/password in the form of Array)
2) added CompilerOptions (in addition to deprecated CompilerOpts) -> I have seen it in some snippets, but no exact syntax
3) added KotlinOptions - see p. 2

Will you support also additional annotations? What about Import/Include? Is there any artifact with annotations definition which can be imported into the project, so that annotations are recognized (like: https://github.com/holgerbrandl/kscript-annotations)?

**Comment by ligee** [comment]
@aartiPl, there are a couple of other annotations, defined in https://github.com/JetBrains/kotlin/blob/master/libraries/tools/kotlin-main-kts/src/org/jetbrains/kotlin/mainKts/annotations.kt, but they are not a part of the "official API" yet, although the underlying mechanisms are part of the API. We may make it official in the future, but it is not decided yet. For us as the API providers, it is important to have the underlying semantics to be common and reused, which annotations are used to trigger it is less important at this point.
But you're free, of course, to reuse the annotations from `main-kts` as well, we just cannot guarantee at the moment that these will stay the same for a long time.

So, to the list of annotations. We have:
- `Repository` probably corresponding to the `MavenRepository` one in `kscript`
- `DependsOn` - corresponding to `DependsOnMaven`
- `CompilerOptions` - from `man-kts`, not part of the API yet
- `Import` - from `man-kts`, not part of the API yet

We do not have anything analogous to `KotlinOpts` as far as I can see, at least in the current script processing model. These annotations processed during script compilation, and default scripting execution model is in-process.
But I can imagine an implementation that takes such an annotation and runs the script with all required JVM opts out-of-process, if needed.

About username/password - you can find more e.g. in this issue - https://youtrack.jetbrains.com/issue/KT-27701#focus=Comments-27-5659673.0-0 and the follow-up.

As of using these things, the best way to my mind to have a look at the `main-kts` implementation https://github.com/JetBrains/kotlin/blob/master/libraries/tools/kotlin-main-kts/src/org/jetbrains/kotlin/mainKts/scriptDef.kt#L45 or simplifide version here - https://github.com/Kotlin/kotlin-script-examples/blob/master/jvm/simple-main-kts/simple-main-kts/src/main/kotlin/org/jetbrains/kotlin/script/examples/simpleMainKts/scriptDef.kt#L49 and implement your version accordingly. 
It is not recommended to use `main-kts` artefact directly, it is designed to be used standalone, and may mess up with your other dependencies.

**Comment by way-zer** [comment]
I have made a [framework](https://github.com/way-zer/ScriptAgent4MindustryExt) using kts as modules. But I encounter a problem, how to using other compiler plugin like `kotlinx servialization` for scripting using annotations.

**Comment by ligee** [comment]
@way-zer, plugins are supported in scripting, but you need to configure it with CLI arguments either via regular CLI compiler or via the `compilerOptions` configuration key.
There is a (quite outdated) test in the Kotlin repo with the CLI compiler usage - https://github.com/JetBrains/kotlin/blob/ca2f37f6ebe9dd7349cc4cab1e1ab1ddfdb5976e/libraries/tools/kotlin-main-kts-test/test/org/jetbrains/kotlin/mainKts/test/mainKtsIT.kt#L145 and https://github.com/JetBrains/kotlin/blob/a7b94b398abc1163ae525ca00c2040e3ad4cc88d/libraries/tools/kotlin-main-kts-test/testData/hello-kotlinx-serialization.main.kts#L1-L0
No IDE support yet unfortunately.

**Comment by way-zer** [comment]
> @way-zer, plugins are supported in scripting, but you need to configure it with CLI arguments either via regular CLI compiler or via the `compilerOptions` configuration key. There is a (quite outdated) test in the Kotlin repo with the CLI compiler usage - https://github.com/JetBrains/kotlin/blob/ca2f37f6ebe9dd7349cc4cab1e1ab1ddfdb5976e/libraries/tools/kotlin-main-kts-test/test/org/jetbrains/kotlin/mainKts/test/mainKtsIT.kt#L145 and https://github.com/JetBrains/kotlin/blob/a7b94b398abc1163ae525ca00c2040e3ad4cc88d/libraries/tools/kotlin-main-kts-test/testData/hello-kotlinx-serialization.main.kts#L1-L0 No IDE support yet unfortunately.

How to use annotation to add plugin. I try CompilerOption:
![Snipaste_2022-09-13_22-42-32](https://user-images.githubusercontent.com/15688938/189931689-508bc785-854b-4ac0-a3df-aa89ddc6721c.png)

**Comment by ligee** [comment]
@way-zer, sorry for the delay.

You're right, the usage via `compilerOptions` is blocked by now, for no good reason. I created an issue about it - https://youtrack.jetbrains.com/issue/KT-54095/It-is-difficult-if-not-impossible-to-use-kotlin-compiler-plugins-with-scripting
will see what we can do about it.

The only way to do it now is to use CLI compiler as in my test.

**Comment by way-zer** [comment]
A new problem: World `refineConfigurationBeforeCompiling` be called when import scripting confuguration in IDE?
It seems not in the latest version(1.7.20), but it did in old version.
So, my library seems not work in new version.

**Comment by gildor** [comment]
@arkanovicz

> easy way of defining an "uber format" that would contain kotlin code inside specific parsed constructs

I think it's totally possible using your own tools which parses any format and extracts Kotlin snippets (for example see the existing (kotlinx-knit)[https://github.com/Kotlin/kotlinx-knit]).
I just don't think that there is a universal solution, for example for Markdown there is standard code block format (triple tick with language name), why one should use some arbitrary template instead?

So, how it can be universal for all possible formats and what is use case for it? One can create a utility that parses those blocks and executes them as programs

Maybe you could clarify your idea, just genuinely interested what do you mean

**Comment by rocketraman** [comment]
> I'm just dreaming aloud of a Kotlin-powered Velocity-like engine...

https://github.com/casid/jte

**Comment by way-zer** [comment]
The `@file:Import` seems support inport pure `kt` source. But it seems not code inspection for the `kt`.
Is there any plan for this function. For now, I am using `kts` in `src` and using `gradle` for dependencies, but which is in deprecated.
I want to know whether is possible to config `dependencies` for imported source,

I am heavily using this feature in [project](https://github.com/way-zer/ScriptAgent4MindustryExt), in which on `kts` is a module, and import some `kt` as lib.

**Comment by ligee** [comment]
@way-zer the importing of `.kt` from script is an unofficial and controversial feature, see https://youtrack.jetbrains.com/issue/KT-28916. We're not sure that it is consistent with the general module concept and worth the support cost.
If you have a compelling use case, please leave a comment in the issue above.

If you're using `.kts` in the sources, please consider using a dedicated definition for it with `isStandalone(false)` property in the configuration.

**Comment by way-zer** [comment]
I'm thinking how to using scripting config instead gradle to config project dependencies. But find it may be impossible.

To be specific, in my project, I'm using `kts` as mirco module, and a script can extend another script classpath. So configuration for one script require another script configuration, but now no refine support that.
May be provide a new `refine` with compileConfig of `importScripts`.
Or may provide another api to get metadata like annotation (just parse without compile).

Another problem, I'd like `importScripts` don't affect `implicitReceivers`, as the usage above(just for compile configuration and classpath). I'm using extenal function to inject the `importScripts` like `contextScipt<Parent>().xxx()`.

**Comment by ligee** [comment]
@way-zer, excuse me, but I cannot grasp the whole problem by your description. Do you have an example code of scripts compilation and refinements, that fails to achieve the results you like?

The refinements are designed with the gradle case in mind, in particular so you should be able to get a callback before compiling of a script, extract a subscript from it, that may affect the classpath and other properties of the main script (e.g. like gradle's `buildScript` section), compile and run it separately to get and updated compilation configuration, and then resume the compilation of the main script.

I do not think that we will provide a separate API for annotations extraction any time soon. This is quite controversial topic and require a lot of discussion and design.

**Comment by way-zer** [comment]
Thank you, I will try the `buildScript` way later.
## Compare `annotation` with `buildScript`
1. `annotation` is simpler, easier, quicker to parse. While `buildScript` more complex (compile, gen class, and eval)
2. `annotation` only support simple type(I test enum before, not know whether works now). `buildScript` can use DSL
3. `annotation` share the `classpath` and `defaultImport`. `buildScript` in IDEA may need more lint (no access outer variable, no access to other library)
### Some little Question
* Can I compile another script in refine callback? That seems in blocking context, while compiling a script needs coroutine context. Is this any problem when in IDE.
* Do I need bundle kotlin-compiler when in IDE, or using the one in kotlin plugin?

## About the `refine`
My usage like this
```kts
// ktor.kts
@Import("io.ktor:ktor-core:xxxx")
@ConfigDefaultImport("io.ktor.http.*")//more like this
//other thing, may init ktor

// b.kts
@Depends("ktor")
// No need to import, as parent config defaultImport
routing {
    get("/") {
        call.respondText("Hello, world!", ContentType.Text.Html)
    }
}
```

In runtime, I can compile `ktor.kts` first, get config for `b.kts`. But in ide, this is not easy to get support.

---

## About `importScripts`
I'm using kts as module, which is the smallest unit to load and unload. Using `contextScript<T>()` for dependency inject. Before, I put all scripts in the same sources when develop, so they can access other's type info.
For example, two module A, B
```kts
// a.kts
fun hello(){
}
// b.kts
@Depends("a")
contextScript<A>().hello()
```
I want to use `importScripts` to compile they together, so `B` can use the class `A`. But I don't like A as the `implicitReceivers` because variable contamination, especially when B has more than one dependency module.

---

## Another Thing
May provide compileOption to assign default package? As above, Multiple modules may have naming conflicts(runtime class conflicts), I want to use directory as package.
For example.
```kts
// a/module.kts
// no need to define package(more like script)

// b.kts
contextScript<a.Module>()

**Comment by way-zer** [comment]
Sorry for disturbing, I'm trying to implement an IDEA plugin to get better support for my framework `ScriptAgent`. But encountered some obstacles.

Two main feature in general:

### Dependency between kts
When `A.kts` depends `B.kts`, A should see the B class (unlike `@file:Import`, A don't get B instance directly).  
**Now**: I'm using  `isStandalone(false)` as a workaround, but too many class in classpath.  
**In runtime**: this is implemented by compiling one by one.(no cycling dependencies)

### Import source(also TODO for LivePlugin)
When `A.kts` import `C.kt`(not `kts`), C will be compiled together with A, also A and B will see each other `class`(in same classpath). `C.kt` here not a inner class of `A.kts`, so it don't need A instance (cycling class dependency, no cycling instance dependency).  
In this way, I could split some of A to `C.kt` or `D.kt` to keep A brief, and in `C.kts` I can expose some API that user don't need instance A to invoke.  
**Now**: I'm using `isStandalone(false)` and `@file:Import(".kt")` as a workaround even they are unofficial and controversial features. 
**In runtime**: use`@file:Import(".kt")` to compile together. (It's better to provide a specialized API to compile `kt` source together)

---

I am trying to implement IDEA plugin to support these features, getting rid of the dependence on unofficial and controversial features. But it seems hard to implement, I have tried `ModuleInfoProviderExtension` with custom `IdeaModuleInfo`, but failled cause to `KotlinCacheServiceImpl` and some internal implement in KT plugin.

So I'd like to know is there any way to resolve the problem, or would you like to give some support?

**Comment by ligee** [comment]
@way-zer, could you, please, ask your questions in the Kotlin Slack? There is a dedicated channel for scripting discussions - https://slack-chats.kotlinlang.org/c/scripting. Alternatively, please create an YouTrack issue, and we can discuss it there. Using this discussion for the particular questions like this is inconvenient.
Just in short - first, you probably need to implement your own variant (or two) of the `@Import` annotation with appropriate handling, so you don't need to hack around the existing `Import` from `.main.kts` and `isStandalone` flag.
Second - there are still missing pieces in the IDE support, namely imported scripts, but we can discuss, what to do with it, e.g. over slack.

**Comment by ligee** [comment]
@jjgp, no, unfortunately there is no progress on that. We may return to this after completely switching to K2.

**Comment by rlconst** [comment]
@ligee K2 is in prod now. Any news regarding kotlin script? Having strongly typed scriptng language is a great thing

**Comment by lare96** [comment]
I use .kts in my [Luna project ](https://github.com/luna-rs/luna/tree/master/src/main/kotlin) heavily as a scripting language. +1 to what @rlconst said, couldn't do what I'm doing now with any other language. 

I really hope scripting support (scripts being compiled and available on cp during runtime with minimal restrictions) is here to stay! Headless statically typed scripts have been such a game changer

**Comment by xGnoSiSx** [comment]
It's been ages, but I'm just here to say that I use scripting support in production, as it is right now, with all the quirks and limitations.

I'd consider it imperative that this gets sorted, solidified, and part of the language libraries and environment. And at these AI perilous times, you can always rewrite the whole thing in another JVM language that properly supports scripting.
